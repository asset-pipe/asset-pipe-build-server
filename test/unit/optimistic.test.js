'use strict';

const Sink = require('@asset-pipe/sink-mem');
const OptimisticBundler = require('../../lib/optimistic-bundler');

beforeAll(() => jest.setTimeout(20000));

const feed1 = [
    {
        id: 'c645cf572a8f5acf8716e4846b408d3b1ca45c58',
        source: '"use strict";module.exports.world=function(){return"world"};',
        deps: {},
        file: './assets/js/bar.js',
    },
];

const feed2 = [
    {
        id: 'b645cf572a8f5acf8716e4846b408d3b1ca45c58',
        source: '"use strict";module.exports.world=function(){return"world"};',
        deps: {},
        file: './assets/js/foo.js',
    },
];

const feed3 = [
    {
        id: 'a645cf572a8f5acf8716e4846b408d3b1ca45c58',
        source: '"use strict";module.exports.world=function(){return"world"};',
        deps: {},
        file: './assets/js/baz.js',
    },
];

test('getFeeds()', async () => {
    const sink = new Sink();
    const ob = new OptimisticBundler({ sink });

    await sink.set('a.json', '1');
    await sink.set('b.json', '2');
    await sink.set('c.json', '3');
    await sink.set('/tags/js/a.txt', 'a');
    await sink.set('/tags/js/b.txt', 'b');
    await sink.set('/tags/js/c.txt', 'c');

    expect(await ob.getFeeds(['a', 'b', 'c'], 'js')).toEqual([1, 2, 3]);
});

test('bundle()', async () => {
    const sink = new Sink();
    const ob = new OptimisticBundler({ sink });

    await sink.set('hash1.json', JSON.stringify(feed1));
    await sink.set('hash2.json', JSON.stringify(feed2));
    await sink.set('hash3.json', JSON.stringify(feed3));
    await sink.set('/tags/js/a.txt', 'hash1');
    await sink.set('/tags/js/b.txt', 'hash2');
    await sink.set('/tags/js/c.txt', 'hash3');

    const buff = [];
    ob.metrics.on('data', chunk => {
        buff.push(chunk);
    });

    const bundle = await ob.bundle(
        'tag',
        ['a', 'b', 'c'],
        ['hash1', 'hash2', 'hash3'],
        'js',
    );
    expect(bundle).toMatchSnapshot();

    ob.metrics.destroy();

    expect(buff).toHaveLength(5);

    const obj = buff[0].toJSON();
    expect(obj.name).toEqual('asset_server_retrieve_from_storage_timer');
    expect(obj.description).toEqual(
        'Time taken for a retrieve operation from storage',
    );
    expect(obj.type).toEqual(5);
    expect(obj.meta).toEqual({ method: 'getFeed' });
});

test('bundleIfNeeded() - produces new bundle', async () => {
    const sink = new Sink();
    const ob = new OptimisticBundler({ sink });

    await sink.set('hash1.json', JSON.stringify(feed1));
    await sink.set('hash2.json', JSON.stringify(feed2));
    await sink.set('hash3.json', JSON.stringify(feed3));
    await sink.set('/tags/js/a.txt', 'hash1');
    await sink.set('/tags/js/b.txt', 'hash2');
    await sink.set('/tags/js/c.txt', 'hash3');

    await ob.bundleIfNeeded({ tag: 'tag', data: ['a', 'b', 'c'], type: 'js' });

    expect(
        await sink.get(
            'e867367ae217ade7ab1acf25afbb04cf3f3ad88cba5022e3c63a328db2124194.js',
        ),
    ).toMatchSnapshot();
});

test('bundleIfNeeded() - produces metrics', async () => {
    const sink = new Sink();
    const ob = new OptimisticBundler({ sink });

    await sink.set('hash1.json', JSON.stringify(feed1));
    await sink.set('hash2.json', JSON.stringify(feed2));
    await sink.set('hash3.json', JSON.stringify(feed3));
    await sink.set('/tags/js/a.txt', 'hash1');
    await sink.set('/tags/js/b.txt', 'hash2');
    await sink.set('/tags/js/c.txt', 'hash3');

    const buff = [];
    ob.metrics.on('data', chunk => {
        buff.push(chunk);
    });

    await ob.bundleIfNeeded({ tag: 'tag', data: ['a', 'b', 'c'], type: 'js' });

    ob.metrics.destroy();

    expect(buff).toHaveLength(8);
    const obj = buff[0].toJSON();
    expect(obj.name).toEqual('asset_server_retrieve_from_storage_timer');
    expect(obj.description).toEqual(
        'Time taken for a retrieve operation from storage',
    );
    expect(obj.type).toEqual(5);
    expect(obj.meta).toEqual({ method: 'getTags' });
});

test('bundleIfNeeded() - does not need to produce bundle', async () => {
    const sink = new Sink();
    const ob = new OptimisticBundler({ sink });

    await sink.set('hash1.json', JSON.stringify(feed1));
    await sink.set('hash2.json', JSON.stringify(feed2));
    await sink.set('hash3.json', JSON.stringify(feed3));
    await sink.set('/tags/js/a.txt', 'hash1');
    await sink.set('/tags/js/b.txt', 'hash2');
    await sink.set('/tags/js/c.txt', 'hash3');

    const file =
        'e867367ae217ade7ab1acf25afbb04cf3f3ad88cba5022e3c63a328db2124194.js';
    await sink.set(file, 'content was not replaced');
    await ob.bundleIfNeeded({ tag: 'tag', data: ['a', 'b', 'c'], type: 'js' });
    expect(
        await sink.get(
            'e867367ae217ade7ab1acf25afbb04cf3f3ad88cba5022e3c63a328db2124194.js',
        ),
    ).toBe('content was not replaced');
});

test('publishAssets() - invalid schema given', async () => {
    const sink = new Sink();
    const ob = new OptimisticBundler({ sink });
    expect(ob.publishAssets({}, 'js')).rejects.toMatchSnapshot();
});

test('publishInstructions() - invalid schema given', async () => {
    const sink = new Sink();
    const ob = new OptimisticBundler({ sink });
    expect(ob.publishInstructions({}, 'js')).rejects.toMatchSnapshot();
});
