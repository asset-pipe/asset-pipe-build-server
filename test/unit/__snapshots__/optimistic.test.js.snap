// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle() 1`] = `
Object {
  "content": "(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({},{},[]);
",
  "hash": "e867367ae217ade7ab1acf25afbb04cf3f3ad88cba5022e3c63a328db2124194",
}
`;

exports[`bundleIfNeeded() - does not need to produce bundle 1`] = `
SinkMem {
  "_events": Object {},
  "_eventsCount": 0,
  "_maxListeners": undefined,
  "db": Object {
    "a-hash1.json": "[{\\"id\\":\\"c645cf572a8f5acf8716e4846b408d3b1ca45c58\\",\\"source\\":\\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",\\"deps\\":{},\\"file\\":\\"./assets/js/bar.js\\"}]",
    "b-hash2.json": "[{\\"id\\":\\"b645cf572a8f5acf8716e4846b408d3b1ca45c58\\",\\"source\\":\\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",\\"deps\\":{},\\"file\\":\\"./assets/js/foo.js\\"}]",
    "c-hash3.json": "[{\\"id\\":\\"a645cf572a8f5acf8716e4846b408d3b1ca45c58\\",\\"source\\":\\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",\\"deps\\":{},\\"file\\":\\"./assets/js/baz.js\\"}]",
    "e867367ae217ade7ab1acf25afbb04cf3f3ad88cba5022e3c63a328db2124194.js": "content was not replaced",
    "meta/tags.json": "{\\"a/js\\":\\"hash1\\",\\"b/js\\":\\"hash2\\",\\"c/js\\":\\"hash3\\"}",
  },
  "domain": null,
  "name": "asset-pipe-sink-mem",
}
`;

exports[`bundleIfNeeded() - produces new bundle 1`] = `
SinkMem {
  "_events": Object {},
  "_eventsCount": 0,
  "_maxListeners": undefined,
  "db": Object {
    "e867367ae217ade7ab1acf25afbb04cf3f3ad88cba5022e3c63a328db2124194.js": "\\"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\\\\\"function\\\\\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\\\\\"Cannot find module '\\\\\\"+o+\\\\\\"'\\\\\\");throw f.code=\\\\\\"MODULE_NOT_FOUND\\\\\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\\\\\"function\\\\\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({},{},[]);\\\\n\\"",
    "hash1.json": "[{\\"id\\":\\"c645cf572a8f5acf8716e4846b408d3b1ca45c58\\",\\"source\\":\\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",\\"deps\\":{},\\"file\\":\\"./assets/js/bar.js\\"}]",
    "hash2.json": "[{\\"id\\":\\"b645cf572a8f5acf8716e4846b408d3b1ca45c58\\",\\"source\\":\\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",\\"deps\\":{},\\"file\\":\\"./assets/js/foo.js\\"}]",
    "hash3.json": "[{\\"id\\":\\"a645cf572a8f5acf8716e4846b408d3b1ca45c58\\",\\"source\\":\\"\\\\\\"use strict\\\\\\";module.exports.world=function(){return\\\\\\"world\\\\\\"};\\",\\"deps\\":{},\\"file\\":\\"./assets/js/baz.js\\"}]",
    "meta/tags.json": "{\\"a/js\\":\\"hash1\\",\\"b/js\\":\\"hash2\\",\\"c/js\\":\\"hash3\\"}",
  },
  "domain": null,
  "name": "asset-pipe-sink-mem",
}
`;

exports[`publishAssets() - invalid schema given 1`] = `
[ValidationError: Invalid 'assets' object given when attempting to publish assets. {
  [41m"tag"[0m[31m [1]: -- missing --[0m
}
[31m
[1] "asset tag" is required[0m]
`;

exports[`publishInstructions() - invalid schema given 1`] = `
[ValidationError: Invalid 'instruction' object given when attempting to publish instructions. {
  [41m"tag"[0m[31m [1]: -- missing --[0m
}
[31m
[1] "asset tag" is required[0m]
`;
